source/source.go

```diff
@@ -127,9 +159,22 @@ func (s *Source) Ack(ctx context.Context, position opencdc.Position) error {
 func (s *Source) Teardown(ctx context.Context) error {
 	sdk.Logger(ctx).Info().Msg("Tearing down the Amazon Redshift Source")
 
-	if s.iterator != nil {
-		if err := s.iterator.Stop(); err != nil {
-			return fmt.Errorf("stop iterator: %w", err)
+	if s.wg != nil {
+		// wait for goroutines to finish
+		s.wg.Wait()
+	}
+
+	if s.ch != nil {
+		// close the read channel for write
+		close(s.ch)
+		// reset read channel to nil, to avoid reading buffered records
+		s.ch = nil
+	}
```

Comment:
IIUC, the wait group will be done only when all iterators are done, and they will be done when they have read all the data, right?

Reply:
Yes, this provides graceful shutdown where all data processing can complete before the connection is closed.

Reply:
Gotcha... But does that mean that if there are 1000 rows in a table, the iterator will wait until it reads all of them?

Reply:
If we have 1000 rows in table and `batchSize` is 100, it will load rows in batches of 100. So if cancel signal is recieved and Teardown() is called, it will wait to finish the loaded batch only.

Reply:
> it will wait to finish the loaded batch only.

meaning: the connector will load the batch from the database, but Conduit will not receive the records? Just trying to make sure we keep Conduit's promise, that Teardown stops reading any new records once it's been called.

Reply:
Okay, I think I got confused earlier. I'll explain the flow:

Suppose we have 1000 rows in table and `batchSize` is 100, the iterator will load first 100 rows and process each row one by one in `next()` and create `sdk.Record`. Once a record is created, either it will be sent to the data channel or if context is cancelled the iterator will stop and the wait group will be done.

`record, err := iter.next(ctx)
		if err != nil {
			sdk.Logger(ctx).Err(err).Msg("iterator shutting down...")
			return //nolint:nlreturn // compact code style
		}

		select {
		case iter.ch <- record:
			position, _ := iter.sourcePosition.get(iter.table)
			iter.position = position

		case <-ctx.Done():
			sdk.Logger(ctx).Debug().Msg("context cancelled, iterator shutting down...")
			err = iter.stop()
			if err != nil {
				sdk.Logger(ctx).Err(err).Msg("failed to properly stop iterator after context cancellation")
			}
			return //nolint:nlreturn // compact code style
		}`
		
So, if 4 records are consumed by conduit and 5th record is ready but context is cancelled it will not push the 5th record to the channel rather it'll shut down the iterator meaning Teardown stops reading any new records once it's been called.